### PR TITLE
Collection of bug fixes: [#3658] [#858] [#3865]

### PR DESCRIPTION
**Pull request by andrew-irods**
Resolves:

#3658 irods/irods ireg should trim trailing slash

In lib/core/src/rodsPath.cpp (around line# 268), a check is made for trailing slashes in the physical directory path, and they are removed if they're there. (This is the code path for dealing with "ireg -C ......").

In server/api/src/rsPhyPathReg.cpp (line# around 730), a check is made by the server for trailing slashes in the physical directory path, and they are removed if they're there before a /filename is appended to that string. This takes care of physical directory paths associated with resources in the catalog pre-dating this fix.

Test case:  test_ireg_double_slashes__issue_3658() in scripts/irods/test/test_ireg.py

Continuous Integration:  Build 713 - http://172.25.14.125:8080/view/2.%20Personal/job/irods-build-and-test-workflow/713/

### CODE REVIEW

**Comment by trel** [comment]
way too many commits - please squash, and re-submit.

consider a single pull request for each issue.

**Comment by andrew-irods** [comment]
Squashed too many commits as requested.
I'm leaving this as a single pull request since the two issues had more than one commit interspersed with each other.  Will keep issues separate in the future.  Thanks.

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,5 @@
 tags
 .dir-locals.el
+.project
+.cproject
+*.swp
```
Comment (trel) [inline .gitignore]: these are unrelated to this issue...

move this edit to a different PR against #858

File: server/api/src/rsPhyPathReg.cpp [hunk 2]
```diff
@@ -727,7 +727,32 @@ dirPathReg( rsComm_t *rsComm, dataObjInp_t *phyPathRegInp, char *filePath,
         fileStatInp_t fileStatInp;
         memset( &fileStatInp, 0, sizeof( fileStatInp ) );
 
-        snprintf( fileStatInp.fileName, MAX_NAME_LEN, "%s/%s", filePath, rodsDirent->d_name );
+        // Issue #3658 - https://github.com/irods/irods/issues/3658
+        //
+        // TODO:  This is a localized fix that addresses any trailing slashes
```
Comment (trel) [inline server/api/src/rsPhyPathReg.cpp]: why TODO?

File: server/api/src/rsPhyPathReg.cpp [hunk 3]
```diff
@@ -727,7 +727,32 @@ dirPathReg( rsComm_t *rsComm, dataObjInp_t *phyPathRegInp, char *filePath,
...
+        // in inPath. Any future refactoring of the code that deals with paths
+        // would most probably take care of this issue, and the code segment
+        // below would/should then be removed.
+        //
+        // This is the second part of a two part fix.  This section removes trailing
```
Comment (trel) [inline server/api/src/rsPhyPathReg.cpp]: reduce comment to just:
`This section removes trailing slashes from the directory path in the server when the path is already in the catalog.`

File: server/api/src/rsPhyPathReg.cpp [hunk 4]
```diff
@@ -727,7 +727,32 @@ dirPathReg( rsComm_t *rsComm, dataObjInp_t *phyPathRegInp, char *filePath,
...
+        rstrcpy( tmpStr, filePath, MAX_NAME_LEN );
+        size_t len = strlen(tmpStr);
+
+        for (size_t i = len-1; i > 0 && tmpStr[i] == '/'; i--) {
+            tmpStr[i] = '\0';
+        }
+
+        snprintf( fileStatInp.fileName, MAX_NAME_LEN, "%s/%s", tmpStr, rodsDirent->d_name );
+
+        //   For debug purposes (issue# 3658)
```
Comment (trel) [inline server/api/src/rsPhyPathReg.cpp]: remove debug code

File: server/api/src/rsPhyPathReg.cpp [hunk 5]
```diff
@@ -727,7 +727,32 @@ dirPathReg( rsComm_t *rsComm, dataObjInp_t *phyPathRegInp, char *filePath,
         fileStatInp_t fileStatInp;
         memset( &fileStatInp, 0, sizeof( fileStatInp ) );
 
-        snprintf( fileStatInp.fileName, MAX_NAME_LEN, "%s/%s", filePath, rodsDirent->d_name );
+        // Issue #3658 - https://github.com/irods/irods/issues/3658
```
Comment (trel) [inline server/api/src/rsPhyPathReg.cpp]: this will be visible due to history - no need to declare in the comment.

File: lib/core/src/rodsPath.cpp [hunk 6]
```diff
@@ -265,6 +265,27 @@ addSrcInPath( rodsPathInp_t *rodsPathInp, const char *inPath ) {
         return USER__NULL_INPUT_ERR;
     }
 
+    // Issue #3658 - https://github.com/irods/irods/issues/3658
```
Comment (trel) [inline lib/core/src/rodsPath.cpp]: entire comment should just be:
`This section removes trailing slashes from the incoming directory path in an invoking icommand (e.g. ireg).`

Comment (andrew-irods) [reply lib/core/src/rodsPath.cpp]: Theses are my opinions:  

The function addSrcInPath() is called from parseCmdLinePath() which is used by a total of nineteen icommands.   Since this removal of a trailing slash from the directory path affects so many utilities, I think it's a good idea to be very verbose here about what's being changed and why. 

I don't think that assuming that the developer looking at this would figure out to go look at the source code control history, or at the git issue detail, is a reasonable assumption. 

In addition, while doing this we (Jason, you, me) discussed the issue of normalizing the treatment of all paths in the future.  We agreed (or I thought we did) that this (removal of trailing slashes in various places in the code) deserves a TODO section that would remind the developers who are doing the normalizing, to clean up (what would then be) unnecessary editing of the path string in various places in the code.  This comment also relates to your comments in rsPhyPathReg.cpp above.

Give me the final verdict, and I'll follow your direction.  Thanks.

**Comment by trel** [comment]
So close - one more pass through rebase to bring up to date.

No need for another Jenkins run.
